lambdas/reconcile-migrated-vcs/src/main/java/uk/gov/di/ipv/core/reconcilemigratedvcs/ReconcileMigratedVcsHandler.java

```diff
@@ -0,0 +1,412 @@
+        gpg45ProfileEvaluator = new Gpg45ProfileEvaluator();
+        tacticalDataStore =
+                DataStore.create(
+                        EnvironmentVariable.USER_ISSUED_CREDENTIALS_TABLE_NAME,
+                        VcStoreItem.class,
+                        configService);
+    }
+
+    @Override
+    public ReconciliationReport handleRequest(Request request, Context context) {
+        reconciliationReport = new ReconciliationReport(request);
+        processedIdentities = new ConcurrentSkipListSet<>();
+        var forkJoinPool =
+                forkJoinPoolFactory.getForkJoinPool(
+                        request.parallelism() == null
+                                ? DEFAULT_PARALLELISM
+                                : request.parallelism());
+        timeoutClose = new AtomicBoolean(false);
+
+        LOGGER.info(
+                LogHelper.buildLogMessage("Starting reconciliation")
+                        .with(LOG_BATCH_ID.getFieldName(), reconciliationReport.getBatchId())
+                        .with("identities", request.userIds().size())
+                        .with("parallelism", forkJoinPool.getParallelism()));
+
+        evcsClient = evcsClientFactory.getClient();
+        evcsClientCount = 1;
+
+        try {
+            forkJoinPool
+                    .submit(
+                            () ->
+                                    request.userIds().parallelStream()
+                                            .unordered()
+                                            .takeWhile(x -> !timeoutClose.get())
+                                            .forEach(userId -> reconcileIdentity(userId, context)))
+                    .get();
+            var exitReason =
+                    timeoutClose.get() ? "Lambda close to timeout" : "All identities processed";
+            LOGGER.info(
+                    LogHelper.buildLogMessage(exitReason)
+                            .with(LOG_BATCH_ID.getFieldName(), reconciliationReport.getBatchId()));
+            reconciliationReport.setExitReason(exitReason);
+        } catch (InterruptedException | ExecutionException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            LOGGER.error(
+                    LogHelper.buildErrorMessage("Parallel execution failed", e)
+                            .with(LOG_BATCH_ID.getFieldName(), reconciliationReport.getBatchId())
+                            .with(LOG_ERROR_STACK.getFieldName(), e.getStackTrace()));
+            reconciliationReport.setExitReason("Parallel execution failed");
+        } finally {
+            forkJoinPool.shutdownNow();
+            var unprocessedIdentities = new ArrayList<>(request.userIds());
+            unprocessedIdentities.removeAll(processedIdentities);
+            reconciliationReport.setUnprocessedHashedUserIds(
+                    unprocessedIdentities.stream().map(DigestUtils::sha256Hex).toList());
```

Comment:
calculating the hash here to avoid returning a list of user identities. This does mean though that we need to do some processing on the result to be able to complete the batch. Do we need to do this, or would returning the unhashed IDs be acceptable - I'm guessing not as we also log the report.

Reply:
Problem is we're logging the report - in case we lose the connection somehow from the local machine. I've got a PR for a script which can do a reverse lookup using the reports table pretty quick, so I'm actually not too worried about it

Reply:
ok sure